crates/fuel-core/src/service/adapters/fuel_gas_price_provider.rs

```diff
@@ -0,0 +1,228 @@
+        Self {
+            totaled_block_height: RefCell::new(block_height),
+            total_reward: RefCell::new(reward),
+            total_cost: RefCell::new(cost),
+        }
+    }
+
+    fn update(&self, block_height: BlockHeight, reward: u64, cost: u64) {
+        let mut totaled_block_height = self.totaled_block_height.borrow_mut();
+        let mut total_reward = self.total_reward.borrow_mut();
+        let mut total_cost = self.total_cost.borrow_mut();
+        *totaled_block_height = block_height;
+        *total_reward += reward;
+        *total_cost += cost;
+    }
+
+    fn totaled_block_height(&self) -> BlockHeight {
+        *self.totaled_block_height.borrow()
+    }
+}
+
+impl<FB, DA, A, GP> FuelGasPriceProvider<FB, DA, A, GP>
+where
+    FB: FuelBlockHistory,
+    DA: DARecordingCostHistory,
+    A: GasPriceAlgorithm,
+    GP: GasPriceHistory,
+{
+    fn inner_gas_price(&self, requested_block_height: BlockHeight) -> Result<u64> {
+        let latest_block = self
+            .block_history
+            .latest_height()
+            .map_err(Error::UnableToGetLatestBlockHeight)?;
+        if latest_block > requested_block_height {
+            let gas_prices = self
+                .gas_price_history
+                .gas_prices(requested_block_height)
+                .map_err(Error::UnableToGetGasPrices)?
+                .ok_or(Error::GasPricesNotFoundForBlockHeight(
+                    requested_block_height,
+                ))?;
+            Ok(gas_prices.total())
+        } else if Self::asking_for_next_block(latest_block, requested_block_height) {
+            let new_gas_prices = self.calculate_new_gas_price(latest_block)?;
+            self.gas_price_history
+                .store_gas_prices(latest_block, &new_gas_prices)
+                .map_err(Error::UnableToStoreGasPrices)?;
+            Ok(new_gas_prices.total())
+        } else {
+            Err(Error::RequestedBlockHeightTooHigh {
+                requested: requested_block_height,
+                latest: latest_block,
+            })
+        }
+    }
+
+    fn asking_for_next_block(
+        latest_block: BlockHeight,
+        block_height: BlockHeight,
+    ) -> bool {
+        *latest_block + 1 == *block_height
+    }
+
+    fn calculate_new_gas_price(
+        &self,
+        latest_block_height: BlockHeight,
+    ) -> Result<GasPrices> {
+        let previous_gas_prices = self
+            .gas_price_history
+            .gas_prices(latest_block_height)
+            .map_err(Error::UnableToGetGasPrices)?
+            .ok_or(Error::GasPricesNotFoundForBlockHeight(latest_block_height))?;
+        let latest_da_recorded_block = self
+            .da_recording_cost_history
+            .latest_height()
+            .map_err(Error::UnableToGetLatestBlockHeight)?;
+
+        let new_da_gas_price = if latest_da_recorded_block
+            <= self.profitablility_totals.totaled_block_height()
+        {
+            previous_gas_prices.da_gas_price
```

Comment:
Since the records are always behind for some period of time, we should always go into this branch. But since `self.profitablility_totals.totaled_block_height()` depends on the number of `self.update_totals` calls, it will not happen. But because of that, it creates very unintuitive code.

What do you think if we always call `self.algorithm.calculate_da_gas_price`, but the `self.total_reward()` and `self.total_cost()` are different? We can achieve almost the same effect just by advancing reward and cost.

Because we always don't have live data from DA, we can predict the cost based on the previous blocks. We can calculate the da gas price per byte based on data from the past(total_cost/total_bytes). Using this price, we can calculate the cost for the upcoming block(txs_size * calculated_gas_price). The `total_cost` is `known_cost_from_the_past` + `cumulutive_predicted_cost`. Using `total_cost`, you can calculate the next gas price, which will give you a reward.

Reply:
Can we know the bytes of the block before we calculate the gas price? I guess theoretically, but the way the code is currently written the gas price is selected before we add the txs to the block, that way we can check that it includes enough to cover fees.

The gas price will never exceed the max estimated gas price, so we can check that each tx has enough to cover that value when adding, then calculate gas price and change after? But that's not how the producer code works rn AFAIK.

We can of course know the older blocks byte size, but I don't know if that helps if we can't compare our current block to it.

Reply:
>Because we always don't have live data from DA, we can predict the cost based on the previous blocks. We can calculate the da gas price per byte based on data from the past(total_cost/total_bytes). Using this price, we can calculate the cost for the upcoming block(txs_size * calculated_gas_price). The total_cost is known_cost_from_the_past + cumulutive_predicted_cost. Using total_cost, you can calculate the next gas price, which will give you a reward.

This is what Optimism does:
>The L1 Data Fee is automatically charged for any transaction that is included in an OP Mainnet block. This fee is deducted directly from the address that sent the transaction. The exact amount paid depends on the estimated size of the transaction in bytes after compression, the current Ethereum gas price and/or blob gas price, and several small parameters.

We don't have the luxury of deducting from the address. And we decided to only have one gas price. So currently we calculate the gas price before we add txs to the block, when we don't know how many bytes the txs will take up in that block. Of course, we could use `dry_run` to calculate the max fee, so that's still an option if we want to separate them.

Reply:
Also relevant from Optimism:
> It is currently not possible to limit the maximum L1 Data Fee that a transaction is willing to pay. This is the result of limitations in existing Ethereum transaction formats and tradeoffs that the OP Stack makes to retain better EVM equivalence. Work is underway to improve this situation with the introduction of a new standardized transaction type. For now, you should be aware that the L1 Data Fee can fluctuate with the Ethereum gas price.